File: SWAIbot Simulator/SWAIbot Simulator Project/Assets/SWAIbot Simulator/Scripts/Protocol/PacketHandler.cs
using UnityEngine;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

public class PacketHandler {
	
	private bool escapeNextChar = false;
	private bool packetReceived = false;
	private StringBuilder buff = new StringBuilder();

	public void received(string str) {
		foreach(char c in str)
			received((int) c);
	}

	public void received(int c) {
		if(c == ((char) '\\')) {
			escapeNextChar = true;
			buff.Append((char) c);
			return;
		}
		
		if(!escapeNextChar) {
			if(c == 1) {
				escapeNextChar = false;
				packetReceived = false;
				buff = new StringBuilder();
				return;
				
			} else if(c == 2) {
				packetReceived = true;
				escapeNextChar = false;
			}
		}
		
		if(packetReceived) {
			Packet p = Protocol.deserialize(buff.ToString());
			escapeNextChar = false;
			packetReceived = false;

			// TODO: Call packet listeners
			/*foreach(PacketListener pl in this.listeners)
				pl.onPacketReceived(p);*/

			SocketController.OnReceived(p);
			
			return;
		}
		
		buff.Append((char) c);
		escapeNextChar = false;
	}
}

File: Eclipse Workspace/SWAIbot/src/com/timvisee/swaibot/protocol/packet/PacketHandler.java
package com.timvisee.swaibot.protocol.packet;

import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;

import com.timvisee.swaibot.protocol.Protocol;

public class PacketHandler {

	private List<PacketListener> listeners = new ArrayList<PacketListener>();
	
	private boolean escapeNextChar = false;
	private boolean packetReceived = false;
	private StringBuilder buff = new StringBuilder();
	
	public static void send(OutputStreamWriter osw, Packet p) {
		try {
			osw.write(Protocol.serialize(p));
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	public void received(int c) {
		if(c == ((char) '\\')) {
			escapeNextChar = true;
			buff.append((char) c);
			return;
		}
		
		if(!escapeNextChar) {
			if(c == 1) {
				escapeNextChar = false;
				packetReceived = false;
				buff = new StringBuilder();
				return;
				
			} else if(c == 2) {
				packetReceived = true;
				escapeNextChar = false;
			}
		}
		
		if(packetReceived) {
			Packet p = Protocol.deserialize(buff.toString());
			escapeNextChar = false;
			packetReceived = false;
			
			for(PacketListener pl : this.listeners)
				pl.onPacketReceived(p);
			
			return;
		}
		
		buff.append((char) c);
		escapeNextChar = false;
	}
	
	public void registerListener(PacketListener pl) {
		this.listeners.add(pl);
	}
	
	public void unregisterListener(PacketListener pl) {
		this.listeners.remove(pl);
	}
}
